backend/persistent/datastores/Athena.js

```diff
@@ -0,0 +1,189 @@
+        });
+    });
+}
+
+/**
+ * Should return a list of tables and their columns that are defined within the database.
+ * @param {object} connection - Connection parameters
+ * @param {string} connection.accessKey - AWS Access Key
+ * @param {string} connection.secretAccessKey - AWS Secret Key
+ * @param {string} connection.region - AWS Region
+ * @param {string} connection.region - AWS Region
+ * @param {string} connection.database - Database name to connect to
+ * @param {string} connection.s3Outputlocation - Location will Athena will output resutls of query
+ * @returns {Promise} that resolves to { columnnames, rows }
+ */
+export function schemas(connection) {
+    const columnnames = ['Table', 'column_name', 'data_type'];
+    const rows = [];
+
+    connection.sqlStatement = `${SHOW_SCHEMA_QUERY} = '${connection.database}'` ;
+    connection.queryTimeout = DEFAULT_QUERY_TIMEOUT;
+    return new Promise(function(resolve, reject) {
+        executeQuery(connection).then(dataSet => {
+            if (dataSet && dataSet.length > 0) {
+                for (let i = 0; i < dataSet.length; i++) {
+                    const data = dataSet[i];
+
+                    if (data && data.Data && data.Data.length > 0) {
+                        if (i !== 0) {
+                            const row = [];
+                            row.push(data.Data[0].VarCharValue); // Table Name
+                            row.push(data.Data[1].VarCharValue); // Column Name
+                            row.push(data.Data[2].VarCharValue); // DataType
+                            rows.push(row);
+                        }
+                    }
+                }
+            }
+            resolve({columnnames, rows});
+        }).catch(err => {
+            Logger.log(`Unexpected Error getting the schemas ${err}`);
+            reject(err);
+        });
+    });
+}
+
+
+/**
+ * Should return a list of tables that are in the database
+ * @param {object} connection - Connection Parameters
+ * @param {string} connection.accessKey - AWS Access Key
+ * @param {string} connection.secretAccessKey - AWS Secret Key
+ * @param {string} connection.region - AWS Region
+ * @param {string} connection.region - AWS Region
+ * @param {string} connection.database - Database name to connect to
+ * @param {string} connection.s3Outputlocation - Location will Athena will output resutls of query
+ * @returns {Promise} that resolves to { columnnames, rows }
+ */
+export function tables(connection) {
+
+    connection.sqlStatement = SHOW_TABLES_QUERY;
+    connection.queryTimeout = DEFAULT_QUERY_TIMEOUT;
+    return new Promise(function(resolve, reject) {
+        executeQuery(connection).then(dataSet => {
+
+            let rst = [];
+            if (dataSet && dataSet.length > 0) {
+                rst = dataSet.map(data => {
+                    if (data && data.Data && data.Data.length > 0) {
+                        return data.Data[0].VarCharValue;
+                    }
```

Comment:
This returns `undefined` for each element that fail `data && data.Data && data.Data.length > 0`.

Is it possible for this test to fail?

Reply:
From the testing that I have done no this should not come back empty.  I updated the code to have it return an empty string as safety message.  This query is returning a list of tables names and should be pretty straight forward.  I ran a bunch of tests and it was always coming back with values.

Reply:
In that case, wouldn't it be better to let it throw an exception? so that the user knows there is a problem with the request.